File: quiz.scala
case class Node(data: Char, var left: Node = null, var right: Node = null)

def serialize(node: Node): String = node match {
  case null => "-"
  case _ => node.data.toString + serialize(node.left) + serialize(node.right)
}

def deserialize(data: String): Node = deserialize2(data.toIterator)

def deserialize2(data: Iterator[Char]): Node = if(data.isEmpty) null else data.next() match {
  case '-' => null
  case a: Char => Node(a, deserialize2(data), deserialize2(data))
}

val root = Node('1')
root.left = Node('2')
root.left.right = Node('3')
root.left.right.left = Node('4')
root.left.right.right=Node('5')

val a1 = serialize(root)
val a2 = serialize(deserialize(a1))

File: quiz.py
class Node:
    def __init__(self, data, left=None, right=None):
        self.left = left
        self.right = right
        self.data = data


def serialize(node):
    if node:
        return str(node.data) + serialize(node.left) + serialize(node.right)
    else:
        return "-"


def deserialize(data):
    return deserialize_ext(iter(list(data)))


def deserialize_ext(data):
    x = next(data, None)
    if not x or x == '-':
        return None
    else:
        return Node(x, deserialize_ext(data), deserialize_ext(data))


a = Node('1')
a.right = Node('2')
a.right.left = Node('3')
a.right.left.left = Node('4')
a.right.left.right = Node('5')

serialized = serialize(a)
print(serialized)

s = serialize(deserialize(serialized))
print(s)



